### DISCUSSION TITLE
Why does `deno info` outputs `/` after `:` in package names

### DISCUSSION THREAD

**Discussion by kuchta** [denoland/deno#26307] [Q&A]
It's not consistent with general use of package names in deno

**Comment by bartlomieju** [comment]
Can you show an example of such output?

**Comment by kuchta** [comment]
Sure

```
: deno info npm:tailwindcss                                                                       10/16/2024 02:58:25 PM
dependencies: 106 unique
size: 14.79MB

npm:/tailwindcss@3.4.14 (5.44MB)
├── npm:/@alloc/quick-lru@5.2.0 (13.77KB)
├── npm:/arg@5.0.2 (13.34KB)
├─┬ npm:/chokidar@3.6.0 (88.09KB)
│ ├─┬ npm:/anymatch@3.1.3 (9.43KB)
│ │ ├── npm:/normalize-path@3.0.0 (9KB)
│ │ └── npm:/picomatch@2.3.1 (87.84KB)
│ ├─┬ npm:/braces@3.0.3 (43.59KB)
│ │ └─┬ npm:/fill-range@7.1.1 (16.35KB)
│ │   └─┬ npm:/to-regex-range@5.0.1 (22.4KB)
│ │     └── npm:/is-number@7.0.0 (9.39KB)
│ ├── npm:/fsevents@2.3.3 (169.16KB)
```

**Reply by bartlomieju** [reply]
So `npm:<package_name>` is a "short-hand" form, that is not really a URL, but `npm:/<package_name>` is a valid URL. We do that conversion internally to not bother users with it, and `deno info` outputs it with slash so you can actually parse them as URLs.

**Reply by kuchta** [reply]
Thank you for clarification, Bartek. But actually both are valid URLs (URIs) parsable by `new URL()` just one with pathname starting with a slash while the other is not which is kind of confusing considering they are not absolute paths, but I think if you generally prefer one form over the other it's better to be consistent in the output with the recommended input (interface)

**Reply by BlackAsLight** [reply]
> Thank you for clarification, Bartek. But actually both are valid URLs (URIs) parsable by `new URL()` just one with pathname starting with a slash while the other is not which is kind of confusing considering they are not absolute paths, but I think if you generally prefer one form over the other it's better to be consistent in the output with the recommended input (interface)

Neither start with a slash. It starts with `npm:/`. What I believe he is saying, like the HTTPS links starts with `http://`, this link starts with `npm:/`.

**Reply by kuchta** [reply]
No, URLs (URIs) don't start with `scheme:/` and neither with `scheme://` unless there's a optional authority (host) section present. See: https://en.wikipedia.org/wiki/URL

<img width="1090" alt="image" src="https://github.com/user-attachments/assets/3cc26be4-08fe-4766-ac0a-4bab7e0dac56">

```
> new URL('npm:/tailwindcss@3.4.14')
URL {
  href: "npm:/tailwindcss@3.4.14",
  protocol: "npm:",
  pathname: "/tailwindcss@3.4.14",
}
> new URL('npm:tailwindcss@3.4.14')
URL {
  href: "npm:tailwindcss@3.4.14",
  protocol: "npm:",
  pathname: "tailwindcss@3.4.14",
}
```

**Reply by 0f-0b** [reply]
If the path of a URL does not start with a slash, it is said that the URL has an opaque path. The ways you can work with such URLs are very limited. For example when resolving a relative URL, if the base URL has an opaque path, the input can only be a `#fragment` or an absolute URL.

```js
new URL("bar", "npm:package/foo").href; // throws. "bar" is neither a fragment or an absolute URL
new URL("bar", "npm:/package/foo").href; // returns "npm:/package/bar"
```

I guess this is why [the change](https://github.com/denoland/deno_semver/pull/6/files#diff-89ebfcfeb012ed523fb1f8ba30649100bbd8444269d867ce726cc646e6789823L500-R494) was made.

**Reply by kuchta** [reply]
Thank you for clarification, @0f-0b ❤️. If you work with it like this, then it makes total sense. But I still think this format shouldn't be part of user interface unless it's used in documentation and unless it's also preferred format on the input side where you have to sanitize it anyway. Should I create an issue for that?
